src/util/parser.js

```diff
@@ -1,21 +1,142 @@
-const crypto = require("crypto");
-
-module.exports = (input) => {
-  const tmpSep = crypto.randomBytes(16).toString('base64');
-  return input
-    // split at "." and before "[", but only if not escaped
-    .replace(/(?<!\\|^)((?:\\{2})*)(?:\.|(?=\[))(?!\[?$)/g, (_, esc) => `${esc}${tmpSep}`)
-    .split(tmpSep)
-    // handle or groups
-    .map((e) => {
-      const orMatches = [];
-      // handle basic or-groups
-      e.replace(/^{(.+)}$/g, (complete, m) => orMatches.push(...m
-        // split at ",", but only if not escaped
-        .replace(/(?<!\\|^)((?:\\{2})*)(?:,)(?!$)/g, (_, esc) => `${esc}${tmpSep}`)
-        .split(tmpSep)));
-      // handle list or-groups (only numbers)
-      e.replace(/^\[{(\d+(?:,\d+)*)}]$/g, (complete, m) => orMatches.push(...m.split(",").map(idx => `[${idx}]`)));
-      return orMatches.length === 0 ? e : orMatches;
-    });
+const assert = require("assert");
+
+const PARENT = Symbol("parent");
+const setParent = (input, parent) => Object.defineProperty(input, PARENT, { value: parent, writable: true });
+const getParent = input => (input[PARENT] === undefined ? null : input[PARENT]);
+
+const OR = Symbol("or");
+const markOr = input => Object.defineProperty(input, OR, { value: true, writable: false });
+const isOr = input => (input[OR] === true);
+module.exports.isOr = isOr;
+
+const Result = (input) => {
```

Comment:
What about `const Parser`? For clarity

Reply:
This encapsulates the result. It doesn't actually do the parsing.